base/nullable.jl

```diff
@@ -247,3 +247,53 @@ function hash(x::Nullable, h::UInt)
         return hash(x.value, h + nullablehash_seed)
     end
 end
+
+# higher-order functions
+"""
+    filter(p, x::Nullable)
+
+Return null if either `x` is null or `p(get(x))` is false, and `x` otherwise.
+"""
+function filter{T}(p, x::Nullable{T})
+    if isbits(T)
+        val = unsafe_get(x)
+        Nullable{T}(val, !isnull(x) && p(val))
+    elseif isnull(x) || p(unsafe_get(x))
+        x
+    else
+        Nullable{T}()
+    end
+end
+
+"""
+Return the given type if it is concrete, and `Union{}` otherwise.
+"""
+concretify{T}(::Type{T}) = isleaftype(T) ? T : Union{}
```

Comment:
Since that's only useful for `Nullable`, and `Union{}` isn't concrete, maybe call this `nullable_returntype` or something like this?